### ISSUE TITLE
Null must be injectable

### ISSUE THREAD

**Issue by rmannibucau** [eclipse/microprofile-config#365]
@Inject
    @ConfigProperty(name="...")
    String value;

It must be allowed to inject that and get null, a workaround is to use an optional but it is not recommanded (even the opposite) to use optionals as fields so this is not a solution. Injecting null is fine from all point of views and compatible with mainstream config solution so it should probably make its path in mp-config asap.

**Comment by Emily-Jiang** [comment]
Do you mean you defined 
`myConfig=`

in config source, you will get empty string instead of null. Do you want to get null instead? If not, can you elaborate your use case?

**Comment by struberg** [comment]
or `optional=true`

**Comment by rmannibucau** [comment]
sounds weird with Optional class no: "optional=true will set null and you can inject Optional<String> which will not be null" :s? But agree a better phrasing can be found.

**Comment by Emily-Jiang** [comment]
-1 on optional=true, as it contradicts with the contract of "the presence of defaultValue" means optional. 

How about to add a new constant on ConfigProperty?
```
String NULL_VALUE="org.eclipse.microprofile.config.configproperty.nullvalue";

@Inject
@ConfigProperty(name="..." defaultValue=ConfigProperty.NULL_VALUE)
String value;
```

**Comment by rmannibucau** [comment]
I had that in mind as well so +1

**Comment by dmlloyd** [comment]
The currently proposed solution does not seem correct.  One alternative discussed in #512 was to add a separate `nullable` annotation property indicating that the injected field should be treated as if it were optional with an `orElse(null)` qualification.  This property would not be allowed to be `true` if the injection target is not nullable (that is, it is primitive).

**Comment by radcortez** [comment]
+1

**Comment by rmannibucau** [comment]
Hmm, why isnt it correct? Having a flag or a flag is the same. One is likely more elegant than the other but both are strictly equivalent.
Now, for the primitive case, you must consider the magic string NULL_VALUE means "the configured value is null but passthrough" so the output takes its default.
Overall point is that duplicating this notion just adds noise to the spec.
Note, however, this is already outside the spec - like geronimo proxy support - since in practise using a default of NULL_VALUE is only valid in cdi injections so not primitives so we are safe anyway.

**Comment by dmlloyd** [comment]
> Hmm, why isnt it correct? Having a flag or a flag is the same. One is likely more elegant than the other but both are strictly equivalent.

They are not strictly equivalent.  How would you then specify the default value for a nullable property?  If default values are implemented using a low priority configuration source (as we do in Quarkus), how do you represent this special behavior?

Bear in mind that we already have a recommended approach for optional values, and that is `Optional`.

> Now, for the primitive case, you must consider the magic string NULL_VALUE means "the configured value is null but passthrough" so the output takes its default.

That is a bit of a stretch, I'd say.  Anyway I believe I already proposed in #476 that primitive types should already implicitly default to their Java default values, which I think is probably a cleaner approach.  It doesn't make these properties optional, but it does remove boilerplate.

> Note, however, this is already outside the spec - like geronimo proxy support - since in practise using a default of NULL_VALUE is only valid in cdi injections so not primitives so we are safe anyway.

I'm not sure what you mean by this.

**Comment by rmannibucau** [comment]
> How would you then specify the default value for a nullable property?

This is the whole point, you can't be both at the same time. If you have a default then you are not nullable so at the end it is just about requesting to accept null as a default. The NULL_VALUE (mp-config 1.4) is exactly there for this case.

> Bear in mind that we already have a recommended approach for optional values, and that is Optional.

As explained before it goes partially against CDI and fully against java recommendation so it can't last IMHO.

The last part which was likely unclear is that the coercing of defaults for primitive is not needed in the spec, there is no way to inject a primitive with CDI so no need to spec how to do it. The only existing case it is useful is the case the impl supports configuration proxies (see an example here https://github.com/apache/geronimo-config/blob/trunk/impl/src/test/java/org/apache/geronimo/config/test/internal/ProxyTest.java#L104) but it is out of the spec for now.

**Comment by dmlloyd** [comment]
> > How would you then specify the default value for a nullable property?
> 
> This is the whole point, you can't be both at the same time. If you have a default then you are not nullable so at the end it is just about requesting to accept null as a default. The NULL_VALUE (mp-config 1.4) is exactly there for this case.

This is factually incorrect.  Users can have a property which is optional/emptiable but also has a default value.  If they want to explicitly clear the property, they may do so.  At least, this works fine on SmallRye/Quarkus.

> > Bear in mind that we already have a recommended approach for optional values, and that is Optional.
> 
> As explained before it goes partially against CDI and fully against java recommendation so it can't last IMHO.

This should be tackled in a separate issue, since `Optional` is used extensively by this API for this purpose.  Using this issue as a proxy for that discussion is counter-productive.

> The last part which was likely unclear is that the coercing of defaults for primitive is not needed in the spec, there is no way to inject a primitive with CDI so no need to spec how to do it. The only existing case it is useful is the case the impl supports configuration proxies (see an example here https://github.com/apache/geronimo-config/blob/trunk/impl/src/test/java/org/apache/geronimo/config/test/internal/ProxyTest.java#L104) but it is out of the spec for now.

OK.

**Comment by rmannibucau** [comment]
>  Users can have a property which is optional/emptiable but also has a default value. 

No, the meaning of defaultValue is "use that if not set". There is no way to set null anywhere in the configuration (getValue in a ConfigSource does not return an optional so no way to know it is null or not set at all).
There is also no reason to set a default value if you support null.
So at the end both cases are exactly the same: what value do you use if you get null from config sources.

What you are speaking about is another topic: how to specify null when there is a default value. In other words it is how to set explicitly null in/from the config source.
This is another topic for the "core" API, not the CDI integration which uses NULL_VALUE as a marker in the @ConfigProperty annotation - as unconfigured marker which is not supported in the getValue/getOptionalValue methods.
I don't think we want to use the constants of @ConfigProperty in Config to use them as marker + in the core API this distinction is not needed because you can set a default using getOptionalValue().orElse  as easily as having a marker.
So at the end, the injection API is covered now, and the standalone one too IMHO.

> Using this issue as a proxy for that discussion is counter-productive.

Not sure I followed this one. Factually Optional fields do not work in passivable CDI beans so it prevents to use the spec and it is where this issue is coming from.

**Comment by emattheis** [comment]
>There is no way to set null anywhere in the configuration (getValue in a ConfigSource does not return an optional so no way to know it is null or not set at all).

This issue has been raised separately and needs to be resolved, IMHO. Personally, I would like to see the empty string `""` used to indicate the presence of a property with an absent value in a config source, as distinct from a `null` which means the property is not present and other sources should be considered.

>The last part which was likely unclear is that the coercing of defaults for primitive is not needed in the spec, there is no way to inject a primitive with CDI so no need to spec how to do it.

Primitives types are valid beans and injectable per CDI. It's true they are not _proxyable_, but that should not be a concern for MicroProfile Config.

**Comment by emattheis** [comment]
The debate about how to indicate nullability in the annotation is largely driving by limitations in the expressiveness of annotation structures. I think the guiding principle here is that we need to avoid further complicating the interpretation of value strings by overloading the use of the `defaultValue` annotation field to convey addition meaning.

It can certainly be argued that this is ambiguous at a glance:
```java
@ConfigProperty(name = "foo", nullable = true, defaultValue = "bar")
```
But it is easy enough to specify that `defaultValue` must be empty when `nullable` is `true`, and CDI gives us a convenient way to detect that condition at bootstrap time and provide meaningful feedback.

**Comment by emattheis** [comment]
Maybe the real issue here is the `defaultValue` support itself. `Config` doesn't have
```java
<T> T getValue(String propertyName, String defaultValue, Class<T> propertyType);
```
So why should we provide such convenience during injection? As recommended in the javadocs, a developer is almost certainly better off simple bundling the default values in `META-INF/microprofile-config.properties`.

**Comment by dmlloyd** [comment]
> The debate about how to indicate nullability in the annotation is largely driving by limitations in the expressiveness of annotation structures. I think the guiding principle here is that we need to avoid further complicating the interpretation of value strings by overloading the use of the `defaultValue` annotation field to convey addition meaning.

+1 here

> It can certainly be argued that this is ambiguous at a glance:
> 
> ```java
> @ConfigProperty(name = "foo", nullable = true, defaultValue = "bar")
> String foo;
> ```
> 
> But so is this:
> 
> ```java
> @ConfigProperty(name = "foo", defaultValue = "bar")
> Optional<String> foo;
> ```
> 
> It is easy enough to specify that `defaultValue` must be empty when `nullable` is `true`, and CDI gives us a convenient way to detect that condition at bootstrap time and provide meaningful feedback.

We support this case.  Having a default is not the same as making a property be required (if you assume the reverse, all kinds of ambiguous situations appear, as previously discussed).

> Maybe the real issue here is the `defaultValue` support itself. `Config` doesn't have
> 
> ```java
> <T> T getValue(String propertyName, String defaultValue, Class<T> propertyType);
> ```

This has been proposed.  But using a config source is more future-proof (calling back to the property expansion discussion).

**Comment by rmannibucau** [comment]
@emattheis the default value is handled in the programmatic API through getOptionalValue. Being programmatic, the dev is responsible for making it efficient and well implemented. Being said the mainstream way to get injections is fields, it must already resolve the defaults for performances reasons. It is also inspired from the 3-4 implementations which created MP-config (deltaspike, sabot and friends) which all ended up doing that cause it was the mainstream case.

**Comment by dmlloyd** [comment]
> @emattheis the default value is handled in the programmatic API through getOptionalValue.

This is just not generally true.

**Comment by emattheis** [comment]
>Having a default is not the same as making a property be required (if you assume the reverse, all kinds of ambiguous situations appear, as previously discussed).

Seems like there are just three cases to worry about for injection:

1. the programmer wants deployment to fail when the property is not configured
2. the programmer wants to be guaranteed a value, even when the property is not configured
3. the programmer is prepared to handle `null` or `Optional`

>This has been proposed. But using a config source is more future-proof (calling back to the property expansion discussion).

Agree. I am certainly not advocating for its introduction, just arguing for `@ConfigProperty` to follow semantically with the `Config` API.

**Comment by emattheis** [comment]
>the default value is handled in the programmatic API through getOptionalValue. Being programmatic, the dev is responsible for making it efficient and well implemented.

Precisely my point. The same option is available post-injection in CDI.

>It is also inspired from the 3-4 implementations which created MP-config (deltaspike, sabot and friends) which all ended up doing that cause it was the mainstream case.

Sometimes following the leader isn't the best way to go 😛. IMHO, hardcoded defaults in source code is an anti-pattern to be avoided.

**Comment by rmannibucau** [comment]
Well it is key to have a standard way to set the default and dont rely on a coded one.
It normalizes the config interaction, to understand the config without digging into any code and enables to generate doc trivially.
It would be a huge regression to be programmatic only and would make the injection almost useless and rewrapped in a custom api as before it exists (if apps dont move back to deltaspike).

**Comment by emattheis** [comment]
>It would be a huge regression to be programmatic only and would make the injection almost useless and rewrapped in a custom api as before it exists (if apps dont move back to deltaspike).

Do you actually have data to support this claim? This is a common theme that comes up around breaking changes. It's tempting to look at an existing API and say that any breaking change would be a large impact, but that assumes users are heavily invested in the feature you are breaking. I'm genuinely curious how we gauge the usage of features and weigh the impact of breaking changes.

That being said, if we removed `defaultValue` from the annotation, anyone using it would simply need to move the value to `META-INF/microprofile-config.properties`. I'm not sure how that would make injection "almost useless" or compel someone to move back to deltaspike.

**Comment by dmlloyd** [comment]
This issue is purely about allowing users to specify that they want a property to be optional without using an explicit `Optional`.

So the only questions that matter are:

* Should we allow users to do this at all?
* Is there any reason that a boolean `nullable` property cannot be supported?
* Is there any problem with making the default value of this property be `false`?

**Comment by radcortez** [comment]
IMO

> * Should we allow users to do this at all?
Yes

> * Is there any reason that a boolean `nullable` property cannot be supported?
No

> * Is there any problem with making the default value of this property be `false`?
No

**Comment by rmannibucau** [comment]
No @dmlloyd, the issue is about enabling null to be injected - as you said - but not to duplicate an existing api. Note that I think NULL_VALUE solves that *already* so this issue is fixed I guess.

Side note: on your question about how do we gauge the usage: i guess we all use our own experience. On my side note a single app wpuld work dropping that feature. Moving to a properties file require new tooling too (for fatjar and living doc for example to cite just 2).

**Comment by emattheis** [comment]
>Note that I think NULL_VALUE solves that already so this issue is fixed I guess.

Only on an unreleased version that is now receiving scrutiny by implementors since we're at RC-2.

My apologies for muddying the waters with the `defaultValue` discussion. @rmannibucau are you firmly opposed to the `nullable` field instead of the sentinel value?

**Comment by rmannibucau** [comment]
Hmm RC1 - not a beta - is released and has NULL_VALUE so guess game is kind of played for this year.

I'm strongly against duplicating any information and/or making it ambiguous.

In other words, nullable=true would be ok if NULL_VALUE is removed since then chain flow is unique (if default is configured else if nullable else fail) IMHO.

**Comment by radcortez** [comment]
I think this is the proposal: replace NULL_VALUE with nullable = true.

**Comment by rmannibucau** [comment]
RC1 and RC2 are final releases (RC are not supposed to have breaking changes) so I'm -1 (as an user so not strictly blocking). Microprofile already has a bad image cause of the breaking changes it does so guess we should learn to limit it as much as possible, in particular when it is just about semantic and not feature related.

**Comment by kenfinnigan** [comment]
@rmannibucau is there a definition that precludes a feature that's present in an RC from being removed in a subsequent RC?

My understanding was that an RC is a possible release candidate for the specification, not a final release. If in evaluating the RC it becomes apparent that a feature has a problem, I don't see it being an issue to remove or adjust a feature between RC releases if necessary.

**Comment by rmannibucau** [comment]
Seems RC is not very well defined (I'm using https://wiki.eclipse.org/MicroProfile/SpecRelease/Release as a ref). My understanding, and how I interpret it is:

1. RC is used cause there is a vote like process @eclipse so it needs some kind of staging before pushing finals, RC are just materializing it
2. RC are released before the final with no structural (breaking) change compared to the final

Overall point is that implementors can always use snapshots to evaluate API changes so guess this definition is fine but also agree it can need to be made more explicit (likely another ticket ;)).

**Comment by kenfinnigan** [comment]
Yeah, it was raised the other day that we haven't voted on spec releases in MP for ages. I think it's been put on hold until the Working Group decision is made because the spec process there would define the voting.

For reference, I recall at least one specification in the past that made breaking changes between RC releases due to issues that came about in verifying an RC.

**Comment by rmannibucau** [comment]
Hmm, I wouldnt say that because it had been done it should be redo.

Let's try another approach: is the 1.4 needed? Can't we go 2.0 directly adding a few other tickets (maybe proxying API to cite one)? There the breaking change would be less an issue I guess. Alternatively it can go in 2.0.

Just to highlight a few points: having specs having breaking changes breaks the full ecosystem in several environments, OSGi is one coming to my mind but it is not the only one. This shouldn't be underestimated because the immediate linking we are thinking about is the classpath because it requires a lot of effort after to catch up for apps (I ignore here the vendor effort but it is real too).

**Comment by radcortez** [comment]
This one was blocking a few things for implementors: https://github.com/eclipse/microprofile-config/issues/390

**Comment by radcortez** [comment]
BTW, if there is no 1.4, there is no breaking change.

**Comment by rmannibucau** [comment]
There are 2 1.4, guess you (= eclipse members) can access the stats on central so see how much it got already consumed to judge more precisely.

**Comment by kenfinnigan** [comment]
I don't agree that it's a breaking change to alter a feature that was only present in the `master` branch previously. My understanding is that the Null behavior wasn't present in 1.3, so whether it is or isn't ultimately in 1.4 it doesn't impact 1.4 being a non-breaking change release.

Are you saying that once an RC is cut there is absolutely no opportunity to say a mistake was made and a feature needs to change? If so, that's not how MicroProfile specifications have been working to date

**Comment by radcortez** [comment]
Even if this is the case, the RC was cut a couple of days ago. To my knowledge, there is no release from any implementor with this version yet, so no end users using it right now.

**Comment by emattheis** [comment]
Also, RC1 and RC2 came on the same day, with nothing prior to that since May 2018 🤔
If there is a plan for alpha or beta releases, I'd love to know.

@rmannibucau I can appreciate your frustration since it looks like you filed this back when 1.3 was released. Personally, I wasn't aware of it until recently.

**Comment by rmannibucau** [comment]
Well, this thread wouldnt have popped out, apache would be releasing a 1.4 version. It is on hold cause of this - and note there is no consumer issue with the snapshot and this issue is cleanly solved too (updated 2 apps with no issue but all null workarounds removed).

If RC are not RC then please 1. call them alpha to remove any ambiguity and add a warning for users that it is unstable and must not be used (what a SNAPSHOT is semantically) 2. write a blog post about it to clearly spread the work cause it is not what RC are commonly.

**Comment by emattheis** [comment]
It seems premature for an implementation to release a version based on a release candidate of the spec. We can argue about what sort of changes can be made during a RC cycle, but the whole point is to have the chance to make them, right?

**Comment by rmannibucau** [comment]
Hmm, dont you mix milestones where, I agree, goal is to test, and RC where goal is to do a final but due to immutable state of releases RC are used in case something unexpectedly bad is found (and then final is just a copy of the first "successful RC") - can be a bad manifest, bad bytecode version and so on, but not a design issue?

**Comment by kenfinnigan** [comment]
It might be that we need to adjust the spec release process to actually do a Beta first to indicate "feature complete" but not candidate release.

In the past, and present, that's been problematic as releases tend to be right before an MP platform release, but it's something we should look at

**Comment by radcortez** [comment]
Anyway, are we reaching a conclusion here? Are we all in agreement to rewrite the feature like @dmlloyd suggested?

**Comment by rmannibucau** [comment]
If all actions are taken at the same time (code change but more importantly making it explicit through the process doc page what RC should be - I don't care you qualify them as "beta" or decide to keep the RC semantic and to do milestones instead + writing somewhere that both RC are "beta" - I let native speakers find more appropriated words) no issue on my side to revert/add nullable.

**Comment by radcortez** [comment]
I think that is reasonable. @Emily-Jiang what do you think?

**Comment by Emily-Jiang** [comment]
I finally caught up with the conversation. 

- As for whether we can make more changes in RCx, yes, we can. In MicroProfile process, we use RC as the experiment for releases. We can have multiple RCs before a final release is cut. Maybe we should align with Jakarta EE process under discussion - Mx - RCx
- as for the issue of specifying null in the default source. If the property does not exist in the configsource, it will be null. We can branch off again to discuss how to represent null.
   

- as for 'nullable = true`, I think it should work. The default value for nullable set to false.

**Comment by Emily-Jiang** [comment]
I thought about this further:
Since we have two ways to specify default value now, we need to either 

- A: disallow the coexist of defaultValue and null 

- B: specify defaultValue has a higher priority over nullable.

`@Inject @ConfigProperty(name="someProp", defaultValue="me", nullable=true) String somePropVaue`
Thoughts?

**Comment by dmlloyd** [comment]
`defaultValue`+`nullable` are just as interoperable as `defaultValue`+`Optional`.  This combination is already supported and meaningful in Quarkus via SmallRye Config.  It is not a question of priority because there should be no conflict, particularly in light of #446: if a user explicitly empties a value, then `null` is injected even if there is a default value specified.

**Comment by rmannibucau** [comment]
@Emily-Jiang hmm, do you assume we keep NULL_VALUE? otherwise it means nullable is only used if unconfigured if the default value I guess so likely your B option. A is not really an option since the annotation has default so both will exists, we can only check the actual value.

However, if we are breaking the API, there is another option: extract defaultValue and nullable as annotation, there is no more ambiguity and startup (build for graalvm friends ;)) time validation will be back with a clean programming model:

    @Inject
    @ConfigProperty
    @DefaultValue("....")
    String myUrl;


    @Inject
    @Nullable
    @ConfigProperty
    String myUrl;

both are valid but

    @Inject
    @ConfigProperty
    @Nullable
    @DefaultValue("....")
    String myUrl;

this last one is not valid

Note that this option also enables to type the default which would be a huge win for a spec on top of CDI:

    @Inject
    @ConfigProperty
    @DefaultValue.String("....")
    String myUrl;



side note: @dmlloyd you mix config injection and value extraction from the injection point and configuration value materialization in Config, this thread is 100% about injection, not about value materialization, maybe you should open another topic?

**Comment by dmlloyd** [comment]
The combination of `nullable` and `Optional` is not meaningful but does not have to be explicitly disallowed (it would not result in incorrect operation, just arguably useless behavior).

**Comment by dmlloyd** [comment]
> @Emily-Jiang hmm, do you assume we keep NULL_VALUE? otherwise it means nullable is only used if unconfigured if the default value I guess so likely your B option. A is not really an option since the annotation has default so both will exists, we can only check the actual value.

No, if you read the thread we're talking about introducing a `nullable` parameter to `@ConfigProperty`.

> However, if we are breaking the API, there is another option: extract defaultValue and nullable as annotation, there is no more ambiguity and startup (build for graalvm friends ;)) time validation will be back with a clean programming model:

We weren't talking about breaking the API, but if you want to propose some of these changes for 2.0 I think that would make sense, at least as a proposal.  I think some of the CDI stuff 

> side note: @dmlloyd you mix config injection and value extraction from the injection point and configuration value materialization in Config, this thread is 100% about injection, not about value materialization, maybe you should open another topic?

You're assuming an injection implementation that is different from what we do.  This isn't a topic, it's a proposal specifically to add a `nullable` attribute, which was pretty well on track with a good degree of consensus to this point.

**Comment by rmannibucau** [comment]
@dmlloyd nullable is not about materializing null for empty strings

> if a user explicitly empties a value, then null is injected even if there is a default value specified

This is typically a bug, it should be an empty string (and it is used in some apps from what I saw, typically to set a prefix/suffix).

So we still have a topic to enable to specify null in the config source, this will likely pass through returning optional<string> instead of String in the source but this is unrelated to the enablement of null injections (default to the injection type default vs explicitly set null).

Wdyt about the proposal of splitting in multiple annotations? It drops all ambiguity and is quite successful in JSON-B for example.

**Comment by dmlloyd** [comment]
> @dmlloyd nullable is not about materializing null for empty strings
> 
> > if a user explicitly empties a value, then null is injected even if there is a default value specified
> 
> This is typically a bug, it should be an empty string (and it is used in some apps from what I saw, typically to set a prefix/suffix).

This is simply not the case universally.  If you maintain this behavior, you create a large number of future problems - inconsistent array and list handling, potential problems relating to the future property expansion proposal, etc.  In Quarkus we've successfully adopted the behavior specified by #446, which allows (for example) patterns like `list.prop=a,${some.prop:},c` to return `[ "a", "c" ]` when `some.prop` is not set.  Prefixes/suffixes are trivially achievable by way of `Optional` and `prefix.orElse("")` as appropriate.

**Comment by Emily-Jiang** [comment]
>  It is not a question of priority because there should be no conflict, particularly in light of #446: if a user explicitly empties a value, then null is injected even if there is a default value specified.
If a user empties the property, the `null` comes from the configsource. `null` will be used. However, if the config property does not exist in any config sources and the defaultValue and nullable specified, which value should be assigned to the variable?

**Comment by Emily-Jiang** [comment]
@rmannibucau the nullable is similar to your originally suggestion of `nullAllowed` as shown in my code snippet. No additional annotation is introduced.

**Comment by dmlloyd** [comment]
> > It is not a question of priority because there should be no conflict, particularly in light of #446: if a user explicitly empties a value, then null is injected even if there is a default value specified.
> 
> If a user empties the property, the `null` comes from the configsource. `null` will be used. However, if the config property does not exist in any config sources and the defaultValue and nullable specified, which value should be assigned to the variable?

If the property does not exist in any config sources, then the default value would be assigned to the variable.

**Comment by Emily-Jiang** [comment]
To clarify:

- `@Config @ConfigProperty(name="some", defaultValue="blah", nullable=true) String figure;`
if the property `some` does not exist in any config source, the value of `figure` will be `blah` not `null`

- `@Config @ConfigProperty(name="some",  nullable=true) String figure;`
if the property `some` does not exist in any config source, the value of `figure` will be `null`
In the spec, since nullable is another way to set defaultValue, we should specify the behaviour as above.

**Comment by emattheis** [comment]
@emilyjiang Assuming `@Config` means `@Inject` above...

As far as which value to use for injection, the following should be equivalent:

```java
@Inject
@ConfigProperty(name="some", defaultValue="blah", nullable=true)
String figure;
```

```java
@Inject
@ConfigProperty(name="some", defaultValue="blah")
Optional<String> figure;
```

You will never get a `null` value or an empty `Optional` injected.

**Comment by radcortez** [comment]
In cases where we are injecting into primitive values like:

```
@ConfigProperty(name="bla", nullable=true)
int bla;
```

This of course, doesn't make much sense. Should we just let the container ignore this or throw a DeploymentException saying that `nullable` is not allowed for primitives?

**Comment by emattheis** [comment]
@radcortez I was just about to comment on that. CDI specifies that `null` injection to primitives results in the default value of the respective type. Not sure if that is a good idea here. I think I'm leaning towards disallowing `nullable=true` when the injection target is a primitive.

**Comment by rmannibucau** [comment]
Nullable means the user value is null and it is ok, injected value is the default of the type, no issue.

**Comment by dmlloyd** [comment]
> To clarify:
> 
> * `@Config @ConfigProperty(name="some", defaultValue="blah", nullable=true) String figure;`
>   if the property `some` does not exist in any config source, the value of `figure` will be `blah` not `null`
> * `@Config @ConfigProperty(name="some",  nullable=true) String figure;`
>   if the property `some` does not exist in any config source, the value of `figure` will be `null`
>   In the spec, since nullable is another way to set defaultValue, we should specify the behaviour as above.

`nullable` must not be considered to be another way to set `defaultValue`, because that eliminates the possibility to support the explicit erasure of property values described in #446.

Under this model, given a property with a default value *and* `nullable=true`, there are three possibilities:

* The property is not present in any configuration source, in which case the default value is used
* The property is explicitly emptied in a configuration source, in which case `null` is used
* The property is explicitly present in a configuration source, in which case the value in that configuration source is used

So as you see, we must consider `nullable` (which is really just a variation on `Optional`) to be orthogonal to default values.

**Comment by rmannibucau** [comment]
No, your second point is a case we dont want null but "". Therefore you just demonstrated it is equivalent to defaultValue I think.

Resetting a value requires a specific value handled in configsources by Config, still nothing linked to this thread.

**Comment by emattheis** [comment]
It's linked to this thread because, ideally, handling of the `defatultValue` string should be the same as a string provided by any other config source. Thus, the motivation for avoiding the concept of a `NULL_VALUE` sentinel until that is straightened out.

**Comment by radcortez** [comment]
> * The property is explicitly emptied in a configuration source, in which case `null` is used

Well, at this point, this doest not apply yet, since there is no way in the API to empty a configuration value. Only here https://github.com/eclipse/microprofile-config/issues/446, right?

**Comment by dmlloyd** [comment]
> No, your second point is a case we dont want null but "". Therefore you just demonstrated it is equivalent to defaultValue I think.

First, I would request that you please not change my words.  We want `null` explicitly in this case; I meant what I said and I said what I meant.

But if you really believe that "" should be returned then you should be busy refuting #446, and also spend some time thinking about what should be returned if the property is an `Integer` or a `List` for example.

> Resetting a value requires a specific value handled in configsources by Config, still nothing linked to this thread.

**Comment by dmlloyd** [comment]
> > * The property is explicitly emptied in a configuration source, in which case `null` is used
> 
> Well, at this point, this doest not apply yet, since there is no way in the API to empty a configuration value. Only here #446, right?

Correct, and in the SmallRye implementation.  Once we have a few more cleanup items accomplished, I will be updating the implementation options for #446 to target 2.0.

**Comment by radcortez** [comment]
Ok, I think we are getting to a conclusion. I've already implemented the changes, but before I've submit, we just need to decide on what to do with `nullable=true` for primitives.

I think it is ok to allow it and use initialise the primitive with their default value.

**Comment by emattheis** [comment]
@dmlloyd I think I glossed over the point @radcortez raised earlier - are you saying we need the ability to distinguish between 'property is not configured' and 'property is explicitly emptied by a config source' when considering whether to use `defaultValue`?

**Comment by dmlloyd** [comment]
> @dmlloyd I think I glossed over the point @radcortez raised earlier - are you saying we need the ability to distinguish between 'property is not configured' and 'property is explicitly emptied by a config source' when considering whether to use `defaultValue`?

Yes, I do, even if only for the purpose of giving space to solve #446.  If we do not do so, then solving #446 will require that this solution be revisited and probably changed.

**Comment by dmlloyd** [comment]
> Ok, I think we are getting to a conclusion. I've already implemented the changes, but before I've submit, we just need to decide on what to do with `nullable=true` for primitives.
> 
> I think it is ok to allow it and use initialise the primitive with their default value.

I think this might muddy the waters further with respect to distinguishing an empty value (which would be forbidden for primitives) from a missing value (which would result in a default value of 0/false to be used for primitives).  I'd rather see `nullable` be forbidden completely for primitives for this reason, and for the simple reason that primitives cannot be `null`, so why should they be allowed to be `nullable`?

**Comment by rmannibucau** [comment]
Let's keep it simple. What we want is:

1. Be able to default an injection to a value
2. Be able to inject the default value of the type and not fail the boot cause there is no configured value
3. (446) be able to reset a set value


As explained, 3 is about config-configsource contract so no link to this issue since this one is not about overriding a value to null but lztting null pass without a failure before, 1 and 2 are solved by NULL_VALUE already (or any of the - at least - 3 other concurrent API: nullable, placeholder, split annotation)

**Comment by emattheis** [comment]
>Yes, I do, even if only for the purpose of giving space to solve #446.

Then I think we should align `defaultValue` behavior with a corresponding `Config` option.

e.g.
```java
<T> getValue(String propertyName, String defaultValue, Class<T> type);
```

I think the CDI producer should be able to delegate directly to the Config API without special access to the underlying implementation.

**Comment by rmannibucau** [comment]
It goes further, as explained you must propagate it to the source with optional return types.
Otherwise +1 for a getValue(ValueRequest) supporting all features of the api - we will add converters at some point, just wait ;).

**Comment by emattheis** [comment]
Yes, but keeping it encapsulated behind `Config` is a good idea, IMHO, otherwise the same rules need to be applied in the CDI layer.

**Comment by radcortez** [comment]
> > Ok, I think we are getting to a conclusion. I've already implemented the changes, but before I've submit, we just need to decide on what to do with `nullable=true` for primitives.
> > I think it is ok to allow it and use initialise the primitive with their default value.
> 
> I think this might muddy the waters further with respect to distinguishing an empty value (which would be forbidden for primitives) from a missing value (which would result in a default value of 0/false to be used for primitives). I'd rather see `nullable` be forbidden completely for primitives for this reason, and for the simple reason that primitives cannot be `null`, so why should they be allowed to be `nullable`?

Nevermind. I got the idea that CDI would inject the default value on a native if the value was `null` via producer or something else. It does not:
https://docs.jboss.org/cdi/spec/1.1.EDR1/html_single/#null

It will just fail. So I guess we can specify that the implementation needs to validate and throw a DeploymentException stating that that combination is not supported. Or, we can just let the CDI implementation handle it and fail anyway.

**Comment by emattheis** [comment]
@radcortez https://docs.jboss.org/cdi/spec/1.2/cdi-spec.html#primitive_types_and_null_values

**Comment by radcortez** [comment]
Ops... that what happens when you search on Google for the spec and don't take the version into account. Silly me!

**Comment by radcortez** [comment]
So, I believe @dmlloyd and @emattheis would prefer to disable `nullable=true` for primitives. I think we should allow it to make it consistent with CDI. Any other thoughts?

**Comment by radcortez** [comment]
`java.lang.OptionalDouble`
`java.lang.OptionalInt`
`java.lang.OptionalLong`

These are the specialised Optional for primitive types. I don't think others exist.

**Comment by emattheis** [comment]
Seems like maybe we're back to square one on this. The original request is really about avoiding the  deployment time validation for non-optional types. Today, that is equivalent to checking for null from the underlying config and whether or not a default is supplied at the injection point. That may not be the case down the road when #446 is resolved.

So, `optional=true` is confusing in light of `Optional<T>` support, and `nullable=true` lead us down this rabbit hole.

How about `required=false` and we sidestep #446 concerns?

By default, `required=true` so this fails at deployment if the property is missing (same as today):
```java
@Inject @ConfigProperty(name="foo") String foo;
```
But this will not (even after #446):
```java
@Inject @ConfigProperty(name="foo", required=false) String foo;
```
And is equivalent from a value perspective to this (today, and post #446):
```java
@Inject @ConfigProperty(name="foo") Optional<String> foo;
```

**Comment by Emily-Jiang** [comment]
`@Inject @ConfigProperty(name="someint", nullable=true) int/Integer some;` if the property does not exist, maybe we should throw DeploymentException as nullable=true is not applicable.

**Comment by Emily-Jiang** [comment]
`java.lang.OptionalDouble
java.lang.OptionalInt
java.lang.OptionalLong

These are the specialised Optional for primitive types. I don't think others exist.`
@radcortez nullable is different from Optional.

**Comment by radcortez** [comment]
> `java.lang.OptionalDouble
> java.lang.OptionalInt
> java.lang.OptionalLong
> 
> These are the specialised Optional for primitive types. I don't think others exist.`
> @radcortez nullable is different from Optional.

Sorry wrong issue, this was meant to #513

**Comment by Emily-Jiang** [comment]
@emattheis `@Inject @ConfigProperty(name="foo") String foo;` means that the property is mandatory. If you add `defaultValue`, it will make it optional.

**Comment by emattheis** [comment]
Sure, but we're back to arguing the semantics of _optional_. Now that I understand the motivations of #446 better, `nullable=true` seems like the worst option 😛 

If optionality dictates whether we validate at deployment time, then `optional=true` makes perfect sense to me despite earlier object in the thread above.

We can just specify that the default is `optional=false` but it will be implicitly set to `true` if the injection target is an optional type.

**Comment by rmannibucau** [comment]
Think @emattheis has a point. Once core API is fixed, CDI layer would be simpler to do cleanly instead of doing the opposite. Let's try to fix our core then come back on this one.

**Comment by radcortez** [comment]
I'm starting to think that we should retract this from 1.4 and actually do a breaking change for 2.0 and allow injection of null by default without any use of additional annotation configuration.

**Comment by rmannibucau** [comment]
I agree Roberto, was an original error IMO, in particular when config can be reloaded these days (cloud etc).

**Comment by dmlloyd** [comment]
That's a fundamental question of "avoid null in the API and CDI usages" i.e. "null is an anti-pattern" versus "avoid Optional in the API and CDI usages in favor of null" i.e. "Optional is an anti-pattern".

This is dangerously close to being a self-defeating, unending discussion in the vein of the "C vs C++" argument that has been raging on usenet for the past 20+ years.  I've been around for long enough to know that neither side is "right" or "wrong".  Traditionally this specification has favored `Optional` for end-user APIs, and I think it makes sense for this reason to stay with this approach (considering also that `Optional` fits in well with the JDK's streaming APIs and other usages, in addition to the historical usage in this specification).

If this specification had historically used `null` instead, then I would have the opposite opinion.  I don't think that hatred for `null` or hatred for `Optional` is sufficient to justify one or the other.  I think that supporting both only adds a layer of confusion.  Nevertheless as long as the proposal remained amicable I for one was willing to entertain it fairly.  But so far the discussion has been quite contentious.  Therefore I think that we should just proceed with the revert and try restarting the discussion later.

The revert is justified because the original change was made without appropriate process, so consensus is (IMO) no more needed for the revert than it was for the original change.

**Comment by dmlloyd** [comment]
> Sure, but we're back to arguing the semantics of _optional_. Now that I understand the motivations of #446 better, `nullable=true` seems like the worst option 😛
> 
> If optionality dictates whether we validate at deployment time, then `optional=true` makes perfect sense to me despite earlier objection in the thread above.
> 
> We can just specify that the default is `optional=false` but it will be implicitly set to `true` if the injection target is an optional type.

Can we at least agree that the semantics of `nullable` (or whatever we call it) would be exactly the same as if the type of the property were `Optional`?  Which is to say, any case where an `Optional` value would be `empty`, a `nullable` property would be `null`?

**Comment by emattheis** [comment]
>Can we at least agree that the semantics of nullable (or whatever we call it) would be exactly the same as if the type of the property were Optional? Which is to say, any case where an Optional value would be empty, a nullable property would be null?

And if so, that goes to your point about whether or not we support both `null` and `Optional` in MicroProfile Config in general. That's certainly the crux of the issue since the original ask is to have the injection equivalent of `Config#getValue` produce `null` instead of throwing an exception at deploy time.

Also, for the record, I don't view this thread as overly contentious and I understand the problem space much better having participated. If my comments seem contentious to other participants, I do apologize, and thanks for everyone for sticking with it!

**Comment by emattheis** [comment]
@rmannibucau I'm curious to know if your motivation for raising the issue was because an injected field was actually going to end up as null, or it was just null at deploy time and you need to bypass the validation?

The latter case is one I'm particularly interested in and it seems that maybe what we really need in the injection layer is not a null-handling solution, but a way to indicate whether deployment time validation is to be performed.

If we don't validate at deploy time, you have until the bean is instantiated to provide the config. Of course, we can't guarantee that `NoSuchElementException` won't occur during injection, but we can't really guarantee that today anyway in the face of mutable sources.

Personally, I advocate for deploy time validation going away entirely.

**Comment by dmlloyd** [comment]
> > Can we at least agree that the semantics of nullable (or whatever we call it) would be exactly the same as if the type of the property were Optional? Which is to say, any case where an Optional value would be empty, a nullable property would be null?
> 
> And if so, that goes to your point about whether or not we support both `null` and `Optional` in MicroProfile Config in general. That's certainly the crux of the issue since the original ask is to have the injection equivalent of `Config#getValue` produce `null` instead of throwing an exception at deploy time.

Precisely.  I contend that we do not need two distinct mechanisms for optionality.  If the problem is deployment time validation, then let's solve that problem.  If the problem is that we really hate `Optional` and regret its usage in MP Config, let's tackle that directly instead.

In other words, let's make absolutely certain that this feature request is not a proxy for some other problem.  It certainly seems like an XY problem the more we discuss it.

Again, either way *we must revert this commit* because it was merged without consensus.

**Comment by rmannibucau** [comment]
I raised this issue for the validation blocker requiring a serious number of config entries to have a defaultValue=fakeDefault instead of just letting null being handled. Optional is not an option cause it is not serializable + half of the time didnt match the expected injection type - libraries signatures and i didnt want  a postconstruct just to unwrap all my conf.

**Comment by dmlloyd** [comment]
Have you tried using `javax.inject.Provider` for your configuration properties?

**Comment by rmannibucau** [comment]
It is not an option since by spec it requires too much work (resolution and reinstantiation, it is as redoing the injection) - no cache - so it is too slow for any runtime having some serious config. But it would have been for an one time startup configuration usage.

**Comment by Emily-Jiang** [comment]
ok. Let's revert the change and do a white board design after Config 1.4 is out! I will move this out of the milestone release and tidy up the release note.

**Comment by Emily-Jiang** [comment]
> Also, for the record, I don't view this thread as overly contentious and I understand the problem space much better having participated. If my comments seem contentious to other participants, I do apologize, and thanks for everyone for sticking with it!

@emattheis no worries. I have not seen anything being contentious. Everyone brings different perspective, which is why collaboration matters. Sometimes, it seems take a long time to get the resolution. I think it worthes it as we covered many angles. Config is a center piece and it is very crucial that we get it right.

**Comment by Emily-Jiang** [comment]
> We can just specify that the default is optional=false but it will be implicitly set to true if the injection target is an optional type.

There is no need to introduce optional=false, as the absence of `defaultValue` means the property is not optional. Otherwise, it is optional.

**Comment by rmannibucau** [comment]
@Emily-Jiang do you remove the validation then and consider defaultValue=unconfigured implies null, or do you keep NULL_VALUE? Not sure i got your last message right.

**Comment by ljnelson** [comment]
In case this is helpful, it would seem to me that CDI-validation-time of configuration injection points is effectively impossible, since [it is (apparently?) deliberately (currently) unspecified whether `ConfigSource`s may come and go at runtime](https://github.com/eclipse/microprofile-config/issues/371#issuecomment-557552345).  So maybe focus on the `null`-versus-`Optional` problem and accept that validation is problematic at best, impossible at worst?

**Comment by rmannibucau** [comment]
Hmm, i think it is specified by the api: configsources are constant for a config instance - we still dont have a way to add/remove it even if a custom source can facade others. However it is also specified that their properties can change and that config is dynamically recalling sources when needed - no cache. Combined with cdi which can recreate bean instances - including appscope ones - it means all is dynamic so validation only makes sense if the app creates a "business" instance at startup without reload.
In such a case, the validation is likely done at that time. Concretely the presence only validation is rarely a good validation (url, query, bounds for ints etc...) so it is revalidated anyway by the runtime.
All that to say that you are right on the validation point which should be dropped.
Now if it is removed the question is: does the required constraint moves to the runtime?
This issue was about to not let it move and let passthrough the "unset" value (default value of the type).
I guess we can maybe "open our mind" to other specs and assumes bean validation would be used in config instances so we drop any validation from here?

**Comment by dmlloyd** [comment]
I think exploring integration with bean validation would be a good idea for 2.0.

**Comment by radcortez** [comment]
Bean Validation is something that should be explored overall in MP, not only with Config, but with other specs, for instances OpenAPI / Bean Validation. I see no reason why we need to duplicate BVal annotations and OpenAPI annotations to declare validations in our openapi document.

**Comment by rmannibucau** [comment]
Side note: bena validation being already cdi integrated it already works ;)

**Comment by emattheis** [comment]
+1 for potentially adopting bean validation as the mechanism, but we still have to decide if injecting to a non-optional type implies `@NotNull`.

I'm still a little unclear on whether the motivation of raising this issue was to avoid validation or to actually support `null` injection.

I think if `null` injection is to be supported, then we must also allow `Config::getValue` to return null instead of throwing an exception.

Per @dmlloyd 's notes above, we don't want to start an `Optional` vs `null` flame war, but we should be clear about MicroProfile Config's stance. The current position implied by the Config API and injection facility, if not formally stated anywhere, is that we're on team `Optional`.

My personal opinion is that low-level APIs like Config should match the lowest common denominator of the platform. Like it or not, Java has `null`. Config should too. Having convenient accessors to avoid excessive `Optional.ofNullable()` boilerplate for those who prefer `Optional`, seems fine, but forcing everyone who prefers `null` to unwrap a throwaway `Optional` seems wasteful.

**Comment by rmannibucau** [comment]
The original issue was just to enable an unconfigured value to end up as null in the injection point. This is why NULL_VALUE was doing the job.
But reworking the getValue() to get a ValueRequest { key, defaultString, canBeNull, overridenConverters, ... } makes sense too to me and would enable to clean up the CDI integration based on a "core" with the same features.

**Comment by dmlloyd** [comment]
> +1 for potentially adopting bean validation as the mechanism, but we still have to decide if injecting to a non-optional type implies `@NotNull`.
> 
> I'm still a little unclear on whether the motivation of raising this issue was to avoid validation or to actually support `null` injection.
> 
> I think if `null` injection is to be supported, then we must also allow `Config::getValue` to return null instead of throwing an exception.
> 
> Per @dmlloyd 's notes above, we don't want to start an `Optional` vs `null` flame war, but we should be clear about MicroProfile Config's stance. The current position implied by the Config API and injection facility, if not formally stated anywhere, is that we're on team `Optional`.

+1, this is clearly the stance of the API.

> My personal opinion is that low-level APIs like Config should match the lowest common denominator of the platform. Like it or not, Java has `null`. Config should too. Having convenient accessors to avoid excessive `Optional.ofNullable()` boilerplate for those who prefer `Optional`, seems fine, but forcing everyone who prefers `null` to unwrap a throwaway `Optional` seems wasteful.

Changing a method that previously never returned `null` to being returning `null` is a pretty nasty compatibility breaker because it will cause previously-working code to start throwing NPEs.  Given the nature of this breakage I don't think it's OK to make such a change, even in a 2.0 release.  So if we do decide to have a variant of `getValue` which can return `null`, it should be a third variant, beyond `getValue` and `getOptionalValue`.

**Comment by dmlloyd** [comment]
BTW my personal opinion is to stick with "team `Optional`".  In the end it really doesn't make that big a difference and it's nice to stay consistent.  I'm no fan of `Optional` but I can honestly say that it hasn't caused a single problem outside the context of CDI.  But I think maybe the CDI problems are more indicative of a systemic issue, which also has other indications (such as the necessity for [things like this](https://github.com/smallrye/smallrye-config/blob/master/implementation/src/main/java/io/smallrye/config/inject/ConfigProducer.java#L47-L150) in order to support injecting basic types - does it really make sense, or should we be injecting configuration object which in turn have properties managed by MP config?).

**Comment by rmannibucau** [comment]
@dmlloyd as a side note, please note this is NOT a breaking change because before the deployment was invalid so the configs are there for all working apps, it will just enable new configs to be null.

**Comment by dmlloyd** [comment]
It is breaking in cases where failing deployment validation is the desired behavior (this would be the definition of "required configuration": such configuration is required and the application should not start up if it is missing).

Today any configuration that is not `Optional` is required.  And before we start up again with the default value argument, adding a default value doesn't *necessarily* make a property optional, it just means there is a value by default, which is a necessary interpretation to clean up the mess that is empty value handling, and again I direct all argument on this point to #446.

**Comment by rmannibucau** [comment]
Hmm, not sure I see your point about:

> It is breaking in cases where failing deployment validation is the desired behavior

Means it is breaking broken cases? Sounds quite acceptable to me.
Concretely there is not real case like that in real life otherwise you could have never deployed your application so this can't create regressions.

That said it brings another argument to remove any validation from the spec: the moment it is validated is more complex than just the startup or instantiation, a whole bunch of code is validated lazily or just not validated in applications due to the libraries layers and condition it encounters. I just met it about a HTTP client, JDBC pool and LDap client configurations.

**Comment by radcortez** [comment]
I can see the argument of a breaking change.

I've seen teams that use this to validate that deployments have everything set in their containers, to force DevOps or admins to set the correct values (you may agree or not with the use case). 

We need to weight things like these before making the decision.

**Comment by dmlloyd** [comment]
> Hmm, not sure I see your point about:
> 
> > It is breaking in cases where failing deployment validation is the desired behavior
> 
> Means it is breaking broken cases? Sounds quite acceptable to me.

You seem to be directly saying that you think that requiring a configuration property is broken behavior.  Is that correct?

> Concretely there is not real case like that in real life otherwise you could have never deployed your application so this can't create regressions.

Given the config may in some cases be external to the application, I believe this is a real possibility, and starting up the application without a required configuration property (leading to possibly undefined behavior) is demonstrably worse than failing the deployment.

> That said it brings another argument to remove any validation from the spec: the moment it is validated is more complex than just the startup or instantiation, a whole bunch of code is validated lazily or just not validated in applications due to the libraries layers and condition it encounters. I just met it about a HTTP client, JDBC pool and LDap client configurations.

OK but surely you must agree that removing all validation from the specification is clearly a bigger discussion?  Essentially you're positing use cases which are substantially different from those which have been posited before.  Those use cases may be important but we can't derail every discussion for incremental improvement because of them, otherwise the spec is essentially dead.

**Comment by rmannibucau** [comment]
> You seem to be directly saying that you think that requiring a configuration property is broken behavior. Is that correct?

No, that today it is indeed correctly configured so that it will not introduce a breaking change to relax this. The feature by itself makes sense and can already be covered by any validation interceptor.

About if the discussion is bigger or not, I'm not sure but it sounds like a good exit door for all this thread to me. No issue on my side if it requests more discussion, PoC or tests on your side. I'm just trying to make MP matching real life use cases without having to use workarounds.

**Comment by dmlloyd** [comment]
> About if the discussion is bigger or not, I'm not sure but it sounds like a good exit door for all this thread to me. No issue on my side if it requests more discussion, PoC or tests on your side. I'm just trying to make MP matching real life use cases without having to use workarounds.

I think we need to collect these use cases, for 2.0 if possible, so I expect this will resurface again soon.

**Comment by emattheis** [comment]
If the ability to inject `null`, to skirt validation or otherwise, is no longer a concern, we should just close this as "works as designed".

Otherwise, I think the non-breaking way forward is a new config method to retrieve potentially null values:
```java
String foo = getNullableValue("foo", String.class);
```
and an equivalent injection mechanism:
```java
@Inject
@Nullable
@ConfigProperty(name="foo")
String foo;
```

In contrast to adding a property to the existing annotation, introducing a new annotation makes it clear that nullability applies only to the injection site, instead of potentially confusing the issue of whether or not to apply a default value when retrieving the config value.

**Comment by rmannibucau** [comment]
Hmm, and then we will need default value handling (to use converter chain) in Config so we'll add getValueWithConverters and getNullableValueWithConverters etc? This is why i proposed a getValue(ValueRequest) method.

I don't know if `@Nullable` is a good name since some tools like findbugs use that simple name to detect things - and they are generally broken with IoC code - but I agree splitting in multiple annotation is more promishing than putting it all in a single one.

**Comment by dmlloyd** [comment]
This discussion seems to have run out of steam, and there are so many comments and directions of the discussion that it is no longer useful for the issue to remain open.

If we want to introduce any ideas from this issue, please either open a new issue labeled `use case` or else link to an existing open issue with that label.
